### PR TITLE
XRDDEV-778 delete cert

### PR DESCRIPTION
**Pull request by jansu76**
https://jira.niis.org/browse/XRDDEV-778

- delete cert
- allowed operations listing and checking for cert and csr

### CODE REVIEW

File: src/proxy-ui-api/src/test/java/org/niis/xroad/restapi/util/TokenTestUtils.java [hunk 1]
```diff
@@ -68,24 +61,63 @@ public static TokenInfo createTestTokenInfo(String friendlyName) {
...
-                true,
-                true,
-                "serial-number",
-                "label",
-                123,
-                TokenStatusInfo.OK,
-                new ArrayList<>(),
-                new HashMap<>());
-        return tokenInfo;
+    public static class TokenInfoBuilder {
```
Comment (carohauta) [inline src/proxy-ui-api/src/test/java/org/niis/xroad/restapi/util/TokenTestUtils.java]: Maybe Lombok's [`@Builder` annotation](https://projectlombok.org/features/Builder) could have been utilized here? I am not 100% sure if it fits this case but if it does it removes a lot of lines from this inner class. Same with [CertificateInfoBuilder](https://github.com/nordic-institute/X-Road/pull/294/files#diff-6a455e99ba4ac4bc2b4a0c614062ad6cR293). Just a thought though

File: src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/converter/KeyConverter.java [hunk 2]
```diff
@@ -43,19 +45,46 @@
...
+    }
+
+    /**
+     * Convert {@link KeyInfo} to openapi {@link Key} object
+     * and populate possibleActions
+     * @param keyInfo
+     */
+    public Key convert(KeyInfo keyInfo, TokenInfo tokenInfo) {
+        if (tokenInfo == null) {
+            throw new NullPointerException("tokenInfo is mandatory to populate possibleActions");
```
Comment (carohauta) [inline src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/converter/KeyConverter.java]: Should this be `IllegalArgumentException` instead? I think seeing a `NullPointerException` could easily be interpreted as a bug in the system.

File: src/common-ui/src/main/java/ee/ria/xroad/commonui/SignerProxy.java [hunk 3]
```diff
@@ -328,6 +334,66 @@ public static CertificateInfo getCertForHash(String hash) throws Exception {
...
+        log.trace("Token and key id with cert request id '{}' found", certRequestId);
+
+        return response;
+    }
+
+    /**
+     * @throws IllegalArgumentException if parameter was not a lowercase string
+     */
+    private static void checkLowerCase(String s) {
+        if (s == null || !s.toLowerCase().equals(s)) {
```
[the remark this answers is not in the document]
Comment (jansu76) [reply src/common-ui/src/main/java/ee/ria/xroad/commonui/SignerProxy.java]: That's a good idea. And there are no other callers of these new SignerProxy methods (getCertForHash, getKeyIdForCerthash, getTokenAndKeyIdForCertHash) so there is no risk of breaking old implementations (which I thought for a moment). I was also a bit worried about consistency, as we don't do this kind of thing on this layer for other fields (like cert.id) but I guess hash has good enough reason to be handled differently.

I think I will do this change.

File: src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/service/StateChangeActionHelper.java [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+    }
+
+    /**
+     * get possible actions for certificate
+     * @param tokenInfo
+     * @param keyInfo
+     * @param certificateInfo
+     * @return
+     */
+    public EnumSet<StateChangeActionEnum> getPossibleCertificateActions(TokenInfo tokenInfo,
```
[the remark this answers is not in the document]
Comment (jansu76) [reply src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/service/StateChangeActionHelper.java]: Yeah, the logic is pretty complex due to the fact it tries to be a carbon copy of what the old codebase was. It might be possible to simplify it somehow, but it's very difficult to say which checks are mandatory and which could possibly be gratuitous - so the new implementation is just a direct copy. Thanks for verifying it!

File: src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/service/TokenCertificateService.java [hunk 5]
```diff
@@ -79,6 +84,7 @@
 public class TokenCertificateService {
```
[the remark this answers is not in the document]
Comment (jansu76) [reply src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/service/TokenCertificateService.java]: Absolutely, excellent suggestion. However I am dreaming about some refactoring to the (Coded) exception handling inside TokenCertificateService.java first. All the `isCausedByNnnn` methods would be shared by CsrService and TokenCertificateService so it is probably easier that way.

I don't have a clear plan on the refactoring yet, but something needs to be done.

File: src/proxy-ui-api/src/main/resources/openapi-definition.yaml [hunk 6]
```diff
@@ -3625,6 +3716,23 @@ components:
...
+    PossibleActions: # ok
+      type: array
+      description: array containing the possible state change actions for this item
+      items:
+        $ref: '#/components/schemas/StateChangeAction'
+    StateChangeAction: # ok
+      type: string
+      format: enum
+      description: action to change state of token, key, cert or csr
+      example: IN_USE
```
[the remark this answers is not in the document]
Comment (jansu76) [reply src/proxy-ui-api/src/main/resources/openapi-definition.yaml]: Good catch. That was actually fixed in https://github.com/nordic-institute/X-Road/pull/306

File: src/proxy-ui-api/src/test/java/org/niis/xroad/restapi/util/TokenTestUtils.java (hunk 1 above)
Comment (jansu76) [reply src/proxy-ui-api/src/test/java/org/niis/xroad/restapi/util/TokenTestUtils.java]: Good idea, I will try if it works here. Always good to remove boilerplate if it's possible. Although Lombok has also it's downsides - but this is tests where we don't need to stress so much about those.

File: src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/converter/KeyConverter.java (hunk 2 above)
Comment (jansu76) [reply src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/converter/KeyConverter.java]: Good question. I researched a bit and `IllegalArgumentException` vs `NullPointerException` for null argument seems to be opinion-based, e.g. https://stackoverflow.com/questions/3881/illegalargumentexception-or-nullpointerexception-for-a-null-parameter/8196334#8196334

Accepted answer speaks for IllegalArgumentException but there are some good arguments for use of NPE, such as use in JDK libraries and Guava. 

I think both are valid choices and neither is clearly wrong, but we should be consistent. I did not find other cases from proxy-ui-api modules, but other modules seem to use IllegalArgumentException, so I will change exception type to that. And let's use that as a convention from now on.

File: src/proxy-ui-api/src/test/java/org/niis/xroad/restapi/util/TokenTestUtils.java (hunk 1 above)
Comment (jansu76) [reply src/proxy-ui-api/src/test/java/org/niis/xroad/restapi/util/TokenTestUtils.java]: I tried this a bit, and I think I am rolling back the attempt. It was close but these details were inconvenient:
- I would rather not change ee.ria.xroad.signer.protocol.dto objects like KeyInfo since a) builder is a test-only concern at least right now, and b) it looks like there's some complexity involved with combining e.g. both @Value and @Builder annotations in the same class https://projectlombok.org/features/Builder - basically, I am hesitant to change those dto objects since that would mean a lot of testing to ensure nothing was broken in other modules
- If dto objects themselves cannot be annotated with @Builder, you would have to annotate a method on TokenTestUtils, and you cannot achieve the same kind of default-value handling as current implementation has that way as far as I can see 

Basically it started to look like too much work for something that would be a positive change, but is not really not broken at the moment. It looked nice though so I will keep that in mind when doing the next builder

```
    @Builder
    public static KeyInfo newKeyInfo(String id,
            String friendlyName,
            KeyUsageInfo keyUsageInfo,
            @Singular List<CertRequestInfo> certRequests) {
            return new KeyInfo(true,
                    keyUsageInfo,
                    friendlyName,
                    id,
                    "label",
                    "public-key",
                    new ArrayList(),
                    certRequests,
                    "sign-mechanism-name");
    }
```

File: src/common-ui/src/main/java/ee/ria/xroad/commonui/SignerProxy.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jansu76) [reply src/common-ui/src/main/java/ee/ria/xroad/commonui/SignerProxy.java]: fixed

File: src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/service/StateChangeActionHelper.java [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+    }
+
+    /**
+     * from keys.js:
+     *             if (cert.cert_deletable &&
+     *                 (cert.cert_saved_to_conf || cert.token_active)) {
+     *                 $("#delete").enable();
+     *             }
+     */
+    private boolean canDelete(TokenInfo tokenInfo, KeyInfo keyInfo,
```
[the remark this answers is not in the document]
Comment (jansu76) [reply src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/service/StateChangeActionHelper.java]: fixed

File: src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/openapi/TokenCertificatesApiController.java [hunk 8]
```diff
@@ -108,13 +119,43 @@ public TokenCertificatesApiController(TokenCertificateService tokenCertificateSe
                 | TokenCertificateService.InvalidCertificateException
                 | TokenCertificateService.AuthCertificateNotSupportedException e) {
             throw new BadRequestException(e);
-        } catch (CertificateAlreadyExistsException | TokenCertificateService.CsrNotFoundException e) {
+        } catch (CertificateAlreadyExistsException | CsrNotFoundException e) {
             throw new ConflictException(e);
         } catch (CertificateNotFoundException e) {
             throw new ResourceNotFoundException(e);
+        } catch (ActionNotPossibleException e) {
+            throw new ConflictException(e);
```
[the remark this answers is not in the document]
Comment (jansu76) [reply src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/openapi/TokenCertificatesApiController.java]: fixed

File: src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/converter/KeyConverter.java (hunk 2 above)
Comment (jansu76) [reply src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/converter/KeyConverter.java]: fixed
